README.pod
=pod

=head1 NAME

Spreadsheet::XLSX::Reader::LibXML - Read xlsx spreadsheet files with LibXML

=head1 SYNOPSIS

The following uses the 'TestBook.xlsx' file found in the t/test_files/ folder

	use strict;
	use warnings;
	use Spreadsheet::XLSX::Reader::LibXML;

	my $parser   = Spreadsheet::XLSX::Reader::LibXML->new();
	my $workbook = $parser->parse( 'TestBook.xlsx' );

	if ( !defined $workbook ) {
		die $parser->error(), "\n";
	}

	for my $worksheet ( $workbook->worksheets() ) {

		my ( $row_min, $row_max ) = $worksheet->row_range();
		my ( $col_min, $col_max ) = $worksheet->col_range();

		for my $row ( $row_min .. $row_max ) {
			for my $col ( $col_min .. $col_max ) {

				my $cell = $worksheet->get_cell( $row, $col );
				next unless $cell;

				print "Row, Col    = ($row, $col)\n";
				print "Value       = ", $cell->value(),       "\n";
				print "Unformatted = ", $cell->unformatted(), "\n";
				print "\n";
			}
		}
		last;# In order not to read all sheets
	}

	###########################
	# SYNOPSIS Screen Output
	# 01: Row, Col    = (0, 0)
	# 02: Value       = Category
	# 03: Unformatted = Category
	# 04: 
	# 05: Row, Col    = (0, 1)
	# 06: Value       = Total
	# 07: Unformatted = Total
	# 08: 
	# 09: Row, Col    = (0, 2)
	# 10: Value       = Date
	# 11: Unformatted = Date
	# 12: 
	# 13: Row, Col    = (1, 0)
	# 14: Value       = Red
	# 16: Unformatted = Red
	# 17: 
	# 18: Row, Col    = (1, 1)
	# 19: Value       = 5
	# 20: Unformatted = 5
	# 21: 
	# 22: Row, Col    = (1, 2)
	# 23: Value       = 2017-2-14 #(shows as 2/14/2017 in the sheet)
	# 24: Unformatted = 41318
	# 25: 
	# More intermediate rows ... 
	# 82: 
	# 83: Row, Col    = (6, 2)
	# 84: Value       = 2016-2-6 #(shows as 2/6/2016 in the sheet)
	# 85: Unformatted = 40944
	###########################



=head1 DESCRIPTION

This is another module for parsing Excel 2007+ workbooks.  The goal of this package is 
three fold.  First, as close as possible produce the same output as is visible in an 
excel spreadsheet with exposure to underlying settings from Excel.  Second, adhere as 
close as is reasonable to the L<Spreadsheet::ParseExcel> API where it doesn't conflict 
with the first objective so that less work would be needed to integrate both of them.  
Third, to provide an XLSX sheet parser that is built on L<XML::LibXML>.  The other 
two primary options for XLSX parsing on CPAN use either a one-off XML parser 
(L<Spreadsheet::XLSX>) or L<XML::Twig> (L<Spreadsheet::ParseXLSX>).  In general if 
either of them already work for you without issue then there is no reason to change to this.  
I personally found some bugs and functionality boundaries in both that I wanted to improve 
and by the time I had educated myself enough to make improvement suggestions including 
root causing the bugs to either the XML parser or the reader logic I had written this.

In the process of learning and building I also wrote some additional features for 
this parser that are not found in the L<Spreadsheet::ParseExcel> package.  For instance 
in the L<SYNOPSIS|/SYNOPSIS> the '$parser' and the '$workbook' are actually the same class.  
You could combine both steps by calling new with the 'file_name' attribute called out.  
Afterward it is still possible to call ->error on the instance.  Another improvement 
(From my perspective) is date handling.  This allows simple pluggable custom output format 
that is more flexible than other options.  I leveraged coercions from L<Type::Coercion> 
to do this but anything that follows that general format will work here.  Additionally,
this is a L<Moose> based package.  As such it is designed to be (fairly) extensible by 
writing roles and adding them to this package rather than requireing that you 
extend the package to some new branch.  Read the full documentation for all opportunities!

On the other hand, L<XML::LibXML> has multiple ways to read an XML file this release only 
has an L<XML::LibXML::Reader> parser option.  Future iterations could include a DOM parser 
option.  Additionally this package does not (yet0 provide the same access to the formatting 
elements provided in L<Spreadsheet::ParseExcel>.  That is on the longish and incomplete TODO 
list.



=head2 Primary Methods

These are the primary ways to use this class.  They can be used to open an .xlsx 
workbook.  They are also the ways to set how the workbook is read.  For information 
on how to retrieve data from the worksheets see the 
L<Worksheet|Spreadsheet::XLSX::Reader::LibXML> documentation.  For additional workbook 
options see the L<Attributes|/Attributes> section.

=head3 new( %attributes )

=over

B<Definition:> This is the way to instantiate an instance of this class.  It can 
accept all the L<Attributes|/Attributes>, some, or none.  If the instance is started with 
no arguments then the L<Methods|/Methods> are needed to open the xlsx file.

B<Accepts:> the L<Attributes|/Attributes>

B<Returns:> An instance of the L<Spreadsheet::XLSX::Reader::LibXML> class

=back

=head3 parse( $file_name, $formatter )

=over

B<Definition:> This is a convenience method to match the L<Spreadsheet::ParseExcel> equivalent.  
It only works if the L<file_name|/file_name> attribute was not set with ->new.  It 
is one way to set the L<file_name|/file_name> and L<default_format_list|/default_format_list>

B<Accepts:>

	$file_name = of a valid xlsx file (required)
	$formatter = see the 'default_format_list' attribute for valid options (optional)

B<Returns:> itself when passing with the xlsx file loaded or undef for failure

=back

=head3 worksheet( $name )

=over

B<Definition:> This method will return an  object to read values in the worksheet.  
If no value is passed to $name then the 'next' worksheet in physical order is 
returned. I<next will NOT wrap>

B<Accepts:> the $name string representing the worksheet object you want to open

B<Returns:> a L<Worksheet|Spreadsheet::XLSX::Reader::LibXML::Worksheet> object with the 
ability to read the worksheet of that name.  Or in 'next' mode it returns undef if 
past the last sheet

B<Example:> using the implied 'next' worksheet;

	while( my $worksheet = $workbook-E<gt>worksheet ){
		# Read the next worksheet
	}

=back

=head3 start_at_the_beginning

=over

B<Definition:> This restarts the 'next' worksheet at the first worksheet

B<Accepts:>nothing

B<Returns:> nothing

=back

=head3 worksheets

=over

B<Definition:> This method will return all the worksheets in the workbook as an array. 
I<Not an array ref>.

B<Accepts:>nothing

B<Returns:> an array of L<Worksheet|Spreadsheet::XLSX::Reader::LibXML::Worksheet> 
objects with all the available worksheets in the array

=back

=head3 worksheet_name( $Int )

=over

B<Definition:> This method returns the worksheet name for a given physical position 
in the worksheet from left to right. It counts from zero even if the workbook is in 
'count_from_one' mode.

B<Accepts:>integers

B<Returns:> the worksheet name

B<Example:> To return only worksheet positions 2 through 4

	for $x (2..4){
		my $worksheet = $workbook->worksheet( $workbook->worksheet_name( $x ) );
		# Read the worksheet here
	}

=back

=head3 worksheet_names

=over

B<Definition:> This method returns an array ref of the worksheet names in the 
workbook.

B<Accepts:>nothing

B<Returns:> an array ref

B<Example:> Another way to parse a workbook without building all the sheets at 
once is;

	for $sheet_name ( @{$workbook->worksheet_names} ){
		my $worksheet = $workbook->worksheet( $sheet_name );
		# Read the worksheet here
	}

=back

=head3 number_of_sheets

=over

B<Definition:> This method returns the count of worksheets in the workbook

B<Accepts:>nothing

B<Returns:> an integer

=back

=head3 error

=over

B<Definition:> This returns the most recent error message logged by the package.  
This method is mostly relevant when an unexpected result is returned by some other 
method.

B<Accepts:>nothing

B<Returns:> an error string.

=back

=head3 get_epoch_year

=over

B<Definition:> This returns the epoch year defined by the worsheet.

B<Accepts:>nothing

B<Returns:> 1900 (= windows) or 1904 (= 1904)

=back

=head2 Attributes

Data passed to new when creating an instance (parser).  For modification of 
these attributes see the listed 'attribute methods'. For more information on 
attributes see L<Moose::Manual::Attributes>.

=head3 error_inst

=over

B<Definition:> This attribute holds an 'error' object instance.  It should have 
several methods for managing errors.  Currently no error codes or error translation 
options are available but this should make implementation of that easier.

B<Default:> a L<Spreadsheet::XLSX::Reader::LibXML::Error> instance with the 
attributes set as;
	
	( should_warn => 0 )

B<Range:> The minimum list of methods to implement for your own instance is;

	error set_error clear_error set_warnings if_warn
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 get_error_inst

B<Definition:> returns this instance

=head4 error

B<Definition:> Used to get the most recently logged error

=head4 set_error

B<Definition:> used to set a new error string

=head4 clear_error

B<Definition:> used to clear the current error string in this attribute

=head4 set_warnings

B<Definition:> used to turn on or off real time warnings when errors are set

=head4 if_warn

B<Definition:> a method mostly used to extend this package and see if warnings 
should be emitted.

=head3 file_name

=over

B<Definition:> This attribute holds the full file name and path for the 
xlsx file to be parsed.

B<Default> no default - this must be provided to read a file

B<Range> any unincrypted xlsx file that can be opened in Microsoft Excel
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 set_file_name

B<Definition:> change the set file name (this will reboot the workbook instance)

=head4 has_file_name

B<Definition:> this is fundamentally a way to see if the workbook loaded correctly

=head3 file_creator

=over

B<Definition:> This holds the information stored in the Excel Metadata 
for who created the file originally.  B<You shouldn't set this attribute 
yourself.>

B<Default> the value from the file

B<Range> A string
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 creator

B<Definition:> returns the name of the file creator

=head3 file_date_created

=over

B<Definition:> This holds the created date in the Excel Metadata 
for when the file was first built.  B<You shouldn't set this attribute 
yourself.>

B<Default> the value from the file

B<Range> A timestamp string (ISO ish)
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 date_created

B<Definition:> returns the date the file was created

=head3 file_modified_by

=over

B<Definition:> This holds the information stored in the Excel Metadata 
for who modified the file last.  B<You shouldn't set this attribute 
yourself.>

B<Default> the value from the file

B<Range> A string
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 modified_by

B<Definition:> returns the user name of the person who last modified the file

=head3 file_date_modified

=over

B<Definition:> This holds the last modified date in the Excel Metadata 
for when the file was last changed.  B<You shouldn't set this attribute 
yourself.>

B<Default> the value from the file

B<Range> A timestamp string (ISO ish)
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 date_modified

B<Definition:> returns the date when the file was last modified

=head3 sheet_parser

=over

B<Definition:> This sets the way the .xlsx file is parsed.  For now the only 
choice is 'reader'.

B<Default> 'reader'

B<Range> 'reader'
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 set_parser_type

B<Definition:> the way to change the parser type

=head4 get_parser_type

B<Definition:> returns the currently set parser type

=head3 count_from_zero

=over

B<Definition:> Excel spreadsheets count from 1.  L<Spreadsheet::ParseExcel> 
counts from zero.  This allows you to choose either way.

B<Default> 1

B<Range> 1 = counting from zero like Spreadsheet::ParseExcel, 
0 = Counting from 1 lke Excel
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 counting_from_zero

B<Definition:> a way to check the current attribute setting

=head4 set_count_from_zero

B<Definition:> a way to change the current attribute setting

=head3 file_boundary_flags

=over

B<Definition:> When you request data past the end of a row or past the bottom 
of the data this package can return 'EOR' or 'EOF' to indicate that state.  
This is especially helpful in 'while' loops.  The other option is to return 
'undef'.  This is problematic if some cells in your table are empty which also 
returns undef.

B<Default> 1

B<Range> 1 = return 'EOR' or 'EOF' flags as appropriate,
0 = return undef when requesting a position that is out of bounds
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 boundary_flag_setting

B<Definition:> a way to check the current attribute setting

=head4 change_boundary_flag

B<Definition:> a way to change the current attribute setting

=head3 empty_is_end

=over

B<Definition:> The excel convention is to read the table left to right and top 
to bottom.  Some tables have uneven columns from row to row.  This allows the 
several methods that take 'next' values to wrap after the last element with data 
rather than going to the max column.

B<Default> 0

B<Range> 1 = treat all columns short of the max column for the sheet as being in 
the table, 0 = end each row after the last cell with data rather than going to the 
max sheet column
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 is_empty_the_end

B<Definition:> a way to check the current attribute setting

=head4 set_empty_is_end

B<Definition:> a way to set the current attribute setting

=head3 from_the_edge

=over

B<Definition:> Some data tables start in the top left corner.  Others do not.  I 
don't reccomend that practice but when aquiring data in the wild it is often good 
to adapt.  This attribute sets whether the file reads from the top left edge or from 
the top row with data and starting from the leftmost column with data.

B<Default> 1

B<Range> 1 = treat the top left corner of the sheet even if there is no data in 
the top row or leftmost column, 0 = Set the minimum row and minimum columns to be 
the first row and first column with data
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 set_from_the_edge

B<Definition:> a way to set the current attribute setting

=head3 default_format_list

=over

B<Definition:> This is a departure from L<Spreadsheet::ParseExcel> for two reasons.  
First, it doesn't use the same modules.  Second, this accepts a role with two methods 
where ParseExcel accepts an object instance.

B<Default> Spreadsheet::XLSX::Reader::LibXML::FmtDefault

B<Range> a L<Moose> role with the methods 'get_defined_excel_format' and 
'change_output_encoding' it should be noted that libxml2 which is the underlying code 
for L<XML::LibXML> allways attempts to get the data into perl friendly strings.  That 
means this should only tweak the data on the way out and does not affect the data on the 
way in.
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 get_default_format_list

B<Definition:> a way to check the current attribute setting

=head4 set_default_format_list

B<Definition:> a way to set the current attribute setting

=head3 format_string_parser

=over

B<Definition:> This is the interpreter that turns the excel into a L<Type::Tiny> coercion.  
If you don't like the output or the method you can write your own Moose Role and add it here.

B<Default> Spreadsheet::XLSX::Reader::LibXML::ParseExcelFormatStrings

B<Range> a L<Moose> role with the method 'parse_excel_format_string'
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 get_format_string_parser

B<Definition:> a way to check the current attribute setting

=head4 set_format_string_parser

B<Definition:> a way to set the current attribute setting

=head3 group_return_type

=over

B<Definition:> Traditionally ParseExcel returns a cell object with lots of methods 
to reveal information about the cell.  In reality this is probably not used very much 
so in the interest of simplifying you can get a cell object instance set to the cell 
information.  Or you can just get the raw value in the cell or you can get the cell value 
formatted either the way the sheet specified or the way you specify.  See the 
'custom_formats' attribute for the L<Spreadsheet::XLSX::Reader::LibXML::Worksheet> class 
to insert custom targeted formats for use with the parser.  All empty cells return undef 
no matter what.

B<Default> instance

B<Range> instance = returns a populated L<Spreadsheet::XLSX::Reader::LibXML::Cell> instance,
unformatted = returns the raw value of the cell with no modifications, value = returns just 
the formatted value stored in the excel cell
		
=back

B<attribute methods> Methods provided to adjust this attribute

=head4 get_group_return_type

B<Definition:> a way to check the current attribute setting

=head4 set_group_return_type

B<Definition:> a way to set the current attribute setting

=head1 BUILD / INSTALL from Source

=over
	
B<1.> Download a compressed file with the code
	
B<2.> Extract the code from the compressed file.  If you are using tar this should work:

        tar -zxvf Spreadsheet-XLSX-Reader-LibXML-v0.xx.tar.gz

B<3.> Change (cd) into the extracted directory

=back

(For Windows find what version of make was used to compile your perl)

	perl  -V:make

Then
(for Windows substitute the correct make function (s/make/dmake/g)?)

	>perl Makefile.PL

	>make

	>make test

	>make install # As sudo/root

	>make clean

=head1 SUPPORT

=over

L<github Spreadsheet::XLSX::Reader::LibXML/issues|https://github.com/jandrew/Spreadsheet-XLSX-Reader-LibXML/issues>

=back

=head1 TODO

=over

B<1.> Add a pivot table reader (Not just read the values from the sheet)

B<2.> Add calc chain methods

B<3.> Add more exposure to workbook formatting methods

B<4.> Build a DOM parser alternative for the sheets
(Theoretically faster than the reader but uses more memory)

=back

=head1 AUTHOR

=over

=item Jed Lund

=item jandrew@cpan.org

=back

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

This software is copyrighted (c) 2014 by Jed Lund

=head1 DEPENDENCIES

=over

B<5.010> - (L<perl>)

L<version>

L<Moose>

L<MooseX::StrictConstructor>

L<MooseX::HasDefaults::RO>

L<Archive::Zip>

L<OLE::Storage_Lite>

L<File::Temp>

L<Type::Tiny> - 0.046

L<MooseX::ShortCut::BuildInstance> - 1.026

L<Carp>- cluck

L<XML::LibXML>

L<DateTimeX::Format::Excel>

L<DateTime::Format::Flexible>

=back

=head1 SEE ALSO

=over

L<Spreadsheet::XLSX>

L<Spreadsheet::ParseXLSX>

L<Log::Shiras|https://github.com/jandrew/Log-Shiras>

=back

=cut


Changes
Revision history for Spreadsheet-XLSX-Reader-LibXML

+ New feature.
- Removed/deprecated feature.
! Bug fix.
* Place holder

NOTE: This software is designated as alpha quality until this notice is removed.

v0.1_1 2014-06-06 10:06:00-07:00 America/Los_Angeles

	+ Initial (non-functional) github release 

v0.5_1    2014-11-07 08:23:08-08:00 America/Los_Angeles

	+ Initial near CPAN ready release with a full test suit for cross platform testing
	+ Changed the Cell class to be just an encapulated data holder
	! Fixed the resources links in the META files
	+ Used type support from Type::Coercions to allow for data transformation definitions
	! The documentation is all wrong.  Next release I promise!
	! Reader parsing support only
		- DOM parsing contract dependant or some (possibly distant) time in the future

v0.7_1    2014-11-12 07:41:39-08:00 America/Los_Angeles

	+ POD for the main module complete
	+ Improved POD for the worksheet module

v0.10.1   2014-11-13 05:37:32-08:00 America/Los_Angeles

	+ Finished the Worksheet POD
	+ Completed the Cell POD
	+ Completed the Error POD
	+ Completed the CellToColumnRow POD
	+ Another (but not final0 proofread of the Spreadsheet::XLSX::Reader::LibXML POD
	+ Release with a Beta rev to try and Index on MetaCPAN

v0.10.2   2014-11-13 10:29:27-08:00 America/Los_Angeles

	+ Fixed and updated the POD based on a MetaCPAN read

v0.10.4   2014-11-16 09:35:36-08:00 America/Los_Angeles

	+ Improved CellToColumnRow POD
	+ Improved the top level POD
	+ Improved the Worksheet POD
	! Removed the SVPerlTalk.odp from the package (as bloat)
		Look in the github repository if you want to review it
	+ Added a new POD file for the SharedStrings class (to be universal across Reader and DOM)
	+ Added a new POD file for the Styles class (to be universal across Reader and DOM)
	+ Added a new POD file for the CalcChain class (to be universal across Reader and DOM)

v0.10.6   2014-11-18 16:46:07-08:00 America/Los_Angeles

	! Fixed several errors in the CellToColumnRow POD
	+ Improved the Error POD
	+ More tweaks to the LibXML POD
	! Fixed a precision bug in ParseExcelFormatStrings

v0.12.2   2014-11-26 06:06:30-08:00 America/Los_Angeles

	+ Improved the FmtDefault POD
	+ Improved the LibXML POD
	+ Extracted several elements of cruft from Types
	+ Reworked the ParseExcelFormatStrings to run on Dispatch Tables for improved consistency
		and to hopefully address CPAN testers fails
	+ moved the change_output_encoding out of ParseExcelFormatStrings an into the cell and value 
		calls for more universal implementation.

v0.12.4   2014-11-26 06:45:17-08:00 America/Los_Angeles

	! Removed the regex match within a regex test to pass older perls in test

v0.14.2   2014-11-26 21:52:22-08:00 America/Los_Angeles

	! Fixed a 32bit precision issue

v0.16.2   2014-11-29 16:43:29-08:00 America/Los_Angeles

	+ Some tweaks to the LibXML POD
	! Another re-write of the ParseExcelFormatStrings.pm to try and fix CPAN Testers fails

v0.18.2   2014-11-30 13:54:00-08:00 America/Los_Angeles

	! Although the last update had more passes on CPAN testers this is another update 
		attempting to reach 100% pass rate

v0.20.2   2014-12-02 12:56:46-08:00 America/Los_Angeles

	+ Finished the POD for ParseExcelFormatStrings.pm
	+ Merged UtilFunctions with ParseExcelFormatStrings since there was no reason 
		for them to be separate
	+ Improved the POD for GetCell.pm
	+ made the methods for clearing the row reading start and end column definitions public

v0.20.4   2014-12-03 13:27:23-08:00 America/Los_Angeles

	+ A round of POD scrubbing for POD completed to this point

v0.22.2   2014-12-04 15:46:15-08:00 America/Los_Angeles

	! Fixed a bug in the error reporting for the value method in the cell
	! Fixed a bug where the sheet position wasn't set correctly when calling for worksheet generation by name
		added a test file for this case
	! Improved temp dir cleanup in error situations - with less code
		this inclues a new explicit DEMOLISH at the cell level

v0.24.2   2014-12-09 05:18:13-08:00 America/Los_Angeles

	+ added the empty_return_type attribute so that this package will emulate ParseExcel
		You can still choose to return undef by setting this attribute to 'undef_string'
	! added another fix in the cleanup since it forced a fail when fails weren't needed

v0.26.2   2014-12-11 08:57:26-08:00 America/Los_Angeles

	! Fixed the 'requires missing role' bug (github issue -> missing role?)

v0.28.2   2014-12-13 20:30:26-08:00 America/Los_Angeles

	! Fixed #15 issue in github (Return precedence error)
	- Closed #2 issue in github (Remove the OLE::Storage_lite dependancy)
	! Closes #17 in github (Allow parse to accept either a file name or a file handle)
		+ Added a file_handle attribute to support passing either file handles or file names
		+ Improved the test file 01-types.t to include coercions and error message checking

v0.30.0   2014-12-15 12:51:23-08:00 America/Los_Angeles

	+ Closed #16 in github by moving the role Spreadsheet::XLSX::Reader::LibXML::LogSpace to Log::Shiras::LogSpace
		and moving all the attribute callouts behind ###LogSD comments
	+ Closed #19 in github by expanding the allowed width of the match data reference pointer when checking the 
		answer for specific errors

v0.30.2   2014-12-25 21:25:32-08:00 America/Los_Angeles

	! Closed #20 in github (pull request for undef FileType message specificity)
	+ Closed #3 in github Move Temp::Dir methods to Archive::Zip in github
		This is a big change since it changes some of the extensibility API's
		It also changes the way the initial file is locked.  It used to unzip the file to a 
		temporory directory and then unlock the file.  Now it holds a lock on the file during the whole read.
	! Closed #18 in github.  The temp dir management is very different

Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Read xlsx spreadsheet files with LibXML",
  "AUTHOR" => "Jed Lund <jandrew\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Spreadsheet-XLSX-Reader-LibXML",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Spreadsheet::XLSX::Reader::LibXML",
  "PREREQ_PM" => {
    "Archive::Zip" => 0,
    "Carp" => 0,
    "DateTime::Format::Flexible" => 0,
    "DateTimeX::Format::Excel" => "v0.12.0",
    "File::Temp" => 0,
    "Moose" => 0,
    "Moose::Role" => 0,
    "MooseX::HasDefaults::RO" => 0,
    "MooseX::ShortCut::BuildInstance" => "v1.8.0",
    "MooseX::StrictConstructor" => 0,
    "OLE::Storage_Lite" => 0,
    "Type::Coercion" => 0,
    "Type::Library" => "0.046",
    "Type::Tiny" => 0,
    "Type::Utils" => 0,
    "Types::Standard" => 0,
    "XML::LibXML::Reader" => 0,
    "lib" => 0,
    "strict" => 0,
    "version" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "Data::Dumper" => 0,
    "Spreadsheet::ParseExcel" => 0,
    "Spreadsheet::XLSX::Reader::CellToColumnRow" => 0,
    "Spreadsheet::XLSX::Reader::Error" => 0,
    "Test::Moose" => 0,
    "Test::Most" => 0
  },
  "VERSION" => "v0.10.1",
  "test" => {
    "TESTS" => "t/Spreadsheet/XLSX/Reader/*.t t/Spreadsheet/XLSX/Reader/LibXML/*.t t/Spreadsheet/XLSX/Reader/LibXML/XMLReader/*.t"
  }
);


my %FallbackPrereqs = (
  "Archive::Zip" => 0,
  "Capture::Tiny" => 0,
  "Carp" => 0,
  "Data::Dumper" => 0,
  "DateTime::Format::Flexible" => 0,
  "DateTimeX::Format::Excel" => "v0.12.0",
  "File::Temp" => 0,
  "Moose" => 0,
  "Moose::Role" => 0,
  "MooseX::HasDefaults::RO" => 0,
  "MooseX::ShortCut::BuildInstance" => "v1.8.0",
  "MooseX::StrictConstructor" => 0,
  "OLE::Storage_Lite" => 0,
  "Spreadsheet::ParseExcel" => 0,
  "Spreadsheet::XLSX::Reader::CellToColumnRow" => 0,
  "Spreadsheet::XLSX::Reader::Error" => 0,
  "Test::Moose" => 0,
  "Test::Most" => 0,
  "Type::Coercion" => 0,
  "Type::Library" => "0.046",
  "Type::Tiny" => 0,
  "Type::Utils" => 0,
  "Types::Standard" => 0,
  "XML::LibXML::Reader" => 0,
  "lib" => 0,
  "strict" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



